src/System.IO.Compression/src/System/IO/Compression/PositionPreservingWriteOnlyStreamWrapper.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Compression
+{
+    internal sealed class PositionPreservingWriteOnlyStreamWrapper : Stream
+    {
+        private readonly Stream _stream;
+        private long _position;
+
+        /// <summary>
+        /// Creates a wrapper for write-only (non-readable, non-seekable) streams that keeps track of <see cref="Position"/> and allows it to be read.
+        /// </summary>
+        /// <param name="stream">The underlying stream, which handles all actual writes.</param>
+        public PositionPreservingWriteOnlyStreamWrapper(Stream stream)
+        {
+            _stream = stream;
+        }
+
+        public override bool CanRead => false;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
+
+        public override long Position
+        {
+            get { return _position; }
+            set
+            {
+                throw new NotSupportedException(SR.NotSupported);
+            }
+        }
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            _position += count;
+            _stream.Write(buffer, offset, count);
+        }
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+        {
+            _position += count;
+            return _stream.BeginWrite(buffer, offset, count, callback, state);
+        }
+
+        public override void EndWrite(IAsyncResult asyncResult) => _stream.EndWrite(asyncResult);
+
+        public override void WriteByte(byte value)
+        {
+            _position += 1;
+            _stream.WriteByte(value);
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            _position += count;
```

Comment:
> but what would the correct position be in that case?

We can't know, but that also means we'd be reporting potentially incorrect results from Position, and I don't know enough about the codebase to know what that could mess up.  I'm wondering if on cancellation/failure, we should set _position to a sentinel value (e.g. -1) and change get_Position to do ```if (_position < 0) throw ...; else return _position;```.  It may not be worth it if nothing bad would come of it, but we should understand it better.